pkg/virtctl/set/set_test.go

```diff
@@ -0,0 +1,192 @@
+		})
+
+		It("should fail with invalid CPU count", func() {
+			cmd := clientcmd.NewRepeatableVirtctlCommand("set", vmName, "--cpu=invalid")
+			Expect(cmd()).To(MatchError(ContainSubstring("invalid argument \"invalid\" for \"--cpu\" flag: strconv.ParseUint: parsing \"invalid\": invalid syntax")))
+		})
+
+		It("should fail with zero CPU count", func() {
+			cmd := clientcmd.NewRepeatableVirtctlCommand("set", vmName, "--cpu=0")
+			Expect(cmd()).To(MatchError(ContainSubstring("invalid CPU count: 0; must be greater than 0")))
+		})
+
+		It("should fail with negative CPU count", func() {
+			cmd := clientcmd.NewRepeatableVirtctlCommand("set", vmName, "--cpu=-1")
+			Expect(cmd()).To(MatchError(ContainSubstring("invalid argument \"-1\" for \"--cpu\" flag: strconv.ParseUint: parsing \"-1\": invalid syntax")))
+		})
+
+		It("should fail with invalid memory size", func() {
+			cmd := clientcmd.NewRepeatableVirtctlCommand("set", vmName, "--memory=invalidSize")
+			Expect(cmd()).To(MatchError(ContainSubstring("invalid memory size: invalidSize")))
+		})
+
+		It("should fail with zero memory size", func() {
+			cmd := clientcmd.NewRepeatableVirtctlCommand("set", vmName, "--memory=0")
+			Expect(cmd()).To(MatchError(ContainSubstring("memory size must be greater than zero")))
+		})
+
+		It("should fail with negative memory size", func() {
+			cmd := clientcmd.NewRepeatableVirtctlCommand("set", vmName, "--memory=-1Gi")
+			Expect(cmd()).To(MatchError(ContainSubstring("memory size must be greater than zero")))
+		})
+
+		It("should fail with invalid timeout flag", func() {
+			cmd := clientcmd.NewRepeatableVirtctlCommand("set", vmName, "--cpu=2", "--timeout=invalid")
+			Expect(cmd()).To(MatchError(ContainSubstring("invalid argument \"invalid\" for \"--timeout\" flag")))
+		})
+	})
+
+	Context("Patch CPU and Memory", func() {
+		It("should succeed with valid CPU", func() {
+			vm := kubecli.NewMinimalVM(vmName)
+			vm.ObjectMeta.Namespace = k8smetav1.NamespaceDefault
+			vm.Spec.Template = &v1.VirtualMachineInstanceTemplateSpec{
+				Spec: v1.VirtualMachineInstanceSpec{},
+			}
+			vmInterface.EXPECT().Get(gomock.Any(), vmName, k8smetav1.GetOptions{}).Return(vm, nil).Times(1)
+			vmInterface.EXPECT().Patch(gomock.Any(), vmName, types.JSONPatchType, gomock.Any(), k8smetav1.PatchOptions{}, gomock.Any()).DoAndReturn(
+				func(ctx context.Context, name string, pt types.PatchType, data []byte, options k8smetav1.PatchOptions, subresources ...string) (*v1.VirtualMachine, error) {
+					patches := []map[string]interface{}{}
+					json.Unmarshal(data, &patches)
+					for _, patch := range patches {
+						if patch["op"] == "replace" && patch["path"] == "/spec/template/spec/domain/cpu/sockets" {
+							if vm.Spec.Template.Spec.Domain.CPU == nil {
+								vm.Spec.Template.Spec.Domain.CPU = &v1.CPU{}
+							}
+							vm.Spec.Template.Spec.Domain.CPU.Sockets = expectedCpuSocketsCount
+						}
+					}
+					return vm, nil
+				}).Times(1)
```

Comment:
I'd prefer to drop any additional logic mimicking the API. We just want virtctl do make the right API calls.

Reply:
If you want you could verify virtctl sends the expected Patch request.

Reply:
sure